File: frontend/src/api/candidateStatus.ts
import { useMutation, useQueryClient } from '@tanstack/react-query'

import { apiRequest } from './client'
import type { CandidateStatus } from './types'
import { vacancyKeys } from './vacancies'

/**
 * Тело PUT /candidate-status. PUT заменяет представление целиком, поэтому
 * форма всегда присылает все поля — отсутствующее означает пустое значение.
 */
export interface CandidateStatusInput {
  cover_letter: string
  sent_at: string | null
  interview_stage: string
  hr_contact: string
  interview_record_url: string
  offer_received: boolean
  offered_salary: number | null
  real_salary: number | null
  market_salary_data: string
}

export function useUpsertCandidateStatus(vacancyId: number) {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: (input: CandidateStatusInput) =>
      apiRequest<CandidateStatus>(`/vacancies/${vacancyId}/candidate-status`, {
        method: 'PUT',
        body: input,
      }),
    onSuccess: () => {
      // Сохранение статуса поднимает updated_at вакансии, поэтому обновить нужно
      // и карточку, и список: порядок строк в таблице меняется.
      void queryClient.invalidateQueries({ queryKey: vacancyKeys.all })
    },
  })
}

File: backend/internal/model/candidate_status.go
package model

import "time"

// CandidateStatus — статус прохождения кандидатом вакансии (п. 4.2 ТЗ).
// Связь с вакансией строго 1:1, поэтому на VacancyID стоит уникальный индекс.
type CandidateStatus struct {
	ID        uint `gorm:"primaryKey"`
	VacancyID uint `gorm:"not null;uniqueIndex"`

	CoverLetter        string
	SentAt             *Date
	InterviewStage     string
	HRContact          string
	InterviewRecordURL string
	OfferReceived      bool
	OfferedSalary      *float64
	RealSalary         *float64
	MarketSalaryData   string

	CreatedAt time.Time
	UpdatedAt time.Time

	// Резюме собеседований приходят из внешних систем (п. 4.3 ТЗ).
	// В MVP таблица только создаётся миграцией, API её не трогает.
	InterviewSummaries []InterviewSummary `gorm:"constraint:OnDelete:CASCADE"`
}

// TableName закрепляет имя таблицы из дизайн-документа.
func (CandidateStatus) TableName() string {
	return "candidate_status"
}
